### DISCUSSION TITLE
RFC: Scoped Mutations

### DISCUSSION THREAD

**Discussion by TkDodo** [TanStack/query#7126] [Ideas]
tl;dr: mutations will get a new, opt-in scoping feature. Putting mutations in the same scope will run them in serial rather than in parallel.

## The Problem

All our mutations run in parallel, which is generally a good thing. We can't know what our mutations do or which relation they have to each other.

This is a good default behaviour, however, there are sometimes situations, especially when `offline` mutations are involved, where we'd want to fire mutations in serial. As an example, firing to `/PATCH` requests to the same resource might result in a `ConcurrentModificationException`.

Also, queuing some mutations that alter _the same_ entity might make the in-between update unnecessary. For example, let's say we're moving a card on a kanban board from column A to B, followed by an immediate move from B to C. In that case, the first mutation is likely unnecessary and should be aborted and / or discarded.

## The proposed solution

As a solution, we'd want to introduce `scoping` of mutations. Mutations with the same scope would be treated as being in the same `queue`. This is an `opt-in` feature. When not given a dedicated `scope`, mutations get their own unique `scope` where no other mutation can be in. This will result in the mutation running immediately and in parallel with all other mutations.

There will be no fuzzy matching of scopes - a scope has to fully match an existing one to be treated as "the same".

### Technical implementation

We currently store all mutations as an `Array<Mutation>` in the `MutationCache`:

https://github.com/TanStack/query/blob/4f168d77ec0e947364a920e424c7442b6a7952fa/packages/query-core/src/mutationCache.ts#L85

The idea is to change this structure into:

```
#mutations: Map<string, Array<Mutation<any, any, any, any>>>
```

where the key of the Map is the scope's id. This will create Mutation Arrays that live "next to each other". When a mutation wants to run, it has to check if its scope allows it. If there is already a running mutation in the scope, we will not run the mutation, but put it into `paused` state instead (`status: 'pending', isPaused: true`). Note that paused was, up until now, only used to pause because there was no network connection. However, there is no limitation as to why we pause them - they can be paused for other reasons, too.

When a mutation finishes (either successfully or erroneously), it has to kick of the next mutation in its scope. This makes sure that mutations that were previously paused will then run once their time has come.

### API

Mutations will get a new, optional property:

```
scope?: {
  id: string
}
```

If set, mutations in the same scope will be guaranteed to run in serial. The object notation enables to later allow for different scope behaviours. For example, a scope could be implemented to abort previously running mutations when a new mutation is added, and kick of the last mutation immediately. It could also get a `takeLatest` kind of behaviour where mutations still run serially, but in-between mutations get discarded. In the first iteration, scopes will only get the default behaviour of queuing up mutations and starting the next one in the scope when they are finished.

## Resuming Paused Mutations

While being offline, mutations will still get paused - unless the `networkMode` is set to ignore that. It is therefore _not_ recommended to mix mutations with different `networkMode` settings in the same scope. As a an example, adding a mutation that should always run into a scope where we already have paused mutations (because of being offline) will result in this mutation also being queued up until being online again, because it depends on the previous mutations to finish before it can start.

When resuming paused mutations after having been offline for some time, only the first mutation of each scope will be continued. The other ones need to wait until their predecessor has finished.

This should fix a couple of bugs around `resumePausedMutations`, especially:

- #4896 
- #6825
- a not yet reported issue where adding more mutations while paused mutations are being resumed will lead to out-of-order updates (I found this while developing the trellix clone)

Note that we tried to fix #4896 but that fix actually introduced #6825. It also makes _all_ mutations run in serial now, which isn't the same as when those mutations were executed while being online. This proposal tries to fix this holistically - mutations should behave the same no matter if they were added while being offline or not.

**Comment by hrastnik** [comment]
This sounds good. Only thing I'd suggest is changing the name from `scope` to `queueId`

**Comment by gotnoklu** [comment]
I like this idea of scoping mutations to run them in a queue. However, I agree with @hrastnik on the naming since relating the ID to a queue clearly gives some insight on what's going on.

**Reply by TkDodo** [reply]
Hmm, not so sure about it. Yes, the behavior I have planned now is a queue, but looking at the things we might be able to do with scopes in the future (like removing in-between elements) makes it not act like a normal queue. I'll think a bit about that :)

**Reply by petter** [reply]
People could possibly mix a general term as `scope` with query keys 🤔 Maybe the name could be more closely related to the feature, something along the lines of `serialGroup`?

**Reply by lunelson** [reply]
I'm personally not bothered by the term `scope`; but FWIW as I read it, the common aspect between the different potential behaviours of the 'scope', and the fundamental difference from the current behaviour, is that they are all some form of _serial_ pattern, so the property could also be called `series`?

**Reply by zilahir** [reply]
i do agree with @TkDodo's way of thinking. I'd actually prefer the `scope` but naming is just naming :). Alltogetehr,t his is a great proposal!

**Reply by hrastnik** [reply]
In my opinion the term `queue` would describe the basic functionality that would be most commonly used. The users would have an intuitive understanding of what the feature does. Then, for advanced use cases, the users will look-up answers like "how do remove request from middle of queue in tanstack query" or similar. 

The name should serve the purpose that when the users read it, they intuitively grasp the concept.

**Reply by TkDodo** [reply]
> People could possibly mix a general term as scope with query keys

`mutationKey` also exists, I hope the distinction is clear, as our keys are arrays and they follow the naming `queryKey` and `mutationKey`.

**Reply by gotnoklu** [reply]
@TkDodo, I get it. I for one am excited about scoped mutations! Great stuff you've got going...

**Comment by sudokoi** [comment]
I like the idea. A few questions:
When do we define the scope of the mutations? passed as options to the `useMutation` call or specifically during the `mutate` call? or both?
Allowing the prior adds a direct dependency among the mutations, while allowing the later has the flexibility of running the same mutations either parallelly or serially depending on the usecase. This could potentially eliminate the use of some of the onSuccess callbacks.

**Reply by TkDodo** [reply]
The way I see it, it would be an argumentative to useMutation. But I think technically, we could also allow it when passed to mutate

**Reply by fweinaug** [reply]
@TkDodo allowing it to be passed to mutate would be a great addition. our use-case is calling the mutation in bulk with different ids which is currently tricky.

**Reply by TkDodo** [reply]
it's technically possible but it would be somewhat inconsistent. consider:

```
const { mutate } = useMutation({ mutationFn, onSuccess, scope: { id: '1' } })

mutate(variables, { scope: { id: '2'}, onSuccess: console.log })
```

with this API, `scope` would be overwritten to `'2'`, but for the other options on mutate (onSuccess, onError, onSettled), that's not the case, because the callbacks run additionally to the ones on `useMutation` if the component is still mounted.

so introducing something that's not intuitive is what I'd like to avoid ...

**Reply by GoldenChrysus** [reply]
I'm very interested in this as well as scoping is a great idea and something I could consider as a reason to switch from SWR to TanStack Query. But the current implementation where the scope is defined on `useMutation` makes it completely unusable to me.

Use case:
I have multiple tables where nearly every cell for every row is editable. I don't have full control over the table's individual row components, so I can't do something like have `useMutation` for every row. I would have one `useMutation` for the entire table. It would be nice if mutation operations for a given row/column were sequential while allowing users to freely mutate other, unrelated records simultaneously if they wish. So I would need several independent scopes on the single mutation. It seems like the only thing I could do currently is scope the entire table so every operation on the table is sequential, but that seems less than ideal when you start building up a decent queue of operations when in reality many of the operations are unrelated to one another and could safely run in parallel.

**Reply by TkDodo** [reply]
I'll think about it

**Reply by bmelnychuk** [reply]
Hey @TkDodo, any updates here? we are also very interested in this

**Reply by jbojcic1** [reply]
I have the similar use case and this would be very beneficial

**Reply by jbojcic1** [reply]
not really. I am referring to the ability to be able to set the scope dynamically either by allowing to pass it to mutate fn like suggested above or if that is not intuitive as mentioned above maybe by allowing to define it as a function that takes mutate params as args so I can build the scope id from params

**Reply by jbojcic1** [reply]
@TkDodo would something like what I did [here in my forked repo](https://github.com/jbojcic1/query/pull/1/files) where you can define scope via function make sense?

**Reply by TkDodo** [reply]
@jbojcic1 we dehydrate the `scope` as part of the `Mutation`, so it should be serializable:

https://github.com/TanStack/query/blob/d4b78185b3ced6e3756887d6cb33059233979e46/packages/query-core/src/hydration.ts#L41

**Reply by jbojcic1** [reply]
I implemented an option to pass scope when calling mutate in my fork and it seems to be working fine. If anyone is interested the change can be seen [here](https://github.com/jbojcic1/query/pull/2). Then in my project I am just applying it with bun patch for now.

@TkDodo, since you don't want to introduce inconsistent api, would it make sense that in the next major version upgrade you just remove the current way to provide the scope via useMutation and just make it a part of mutate options? That way you could still have static scope but you would also get the option to make it dynamic which is pretty powerful option imo.

**Reply by TkDodo** [reply]
> I'll think about it

I thought about it and the way I see it, conceptually, we should define one scope for mutation and then call `.mutate()` multiple times on that mutation will be handled by the defined `scope`. The moment you do different scopes per `.mutate` invocation, it _should_ likely become different mutations altogether.

>  I don't have full control over the table's individual row components, so I can't do something like have useMutation for every row.

Sorry to say but multiple `useMutations`, one for each row, is what you want here. Each row should be its own `<Row />` react component that should be able to call `useMutation()` inside. I can only assume that some form of abstraction prevents you from doing that, but I’d rather rethink the abstraction then instead of trying to bend things to fit into this model.

**Reply by jbojcic1** [reply]
My use case is something like this. I am subscribing to some third party data updates via socket. When the record changes the state in that third party system I am notified and then I need to update the state in my Supabase database. To update the state in my database I am using the mutation. If I define the "global" scope, that would make the updates for all the records synchronised. When I provide the scope per mutate call, I get only the updates for the single record synchronised. 
Any suggestion how to handle that? Rendering react components (or creating one MutationObserver if I do it outside react) for each record I am subscribing to seems wrong.

**Reply by TkDodo** [reply]
> When the record changes the state in that third party system I am notified and then I need to update the state in my Supabase database.

Maybe I’m misunderstand but why are you doing this on the frontend at all? It seems like synchronizing two systems that should happen in the backend.

> If I define the "global" scope, that would make the updates for all the records synchronised. When I provide the scope per mutate call, I get only the updates for the single record synchronised.

I think you’d want to look into TanStack Pacer, which has queuing capabilities:

https://tanstack.com/pacer/latest/docs/guides/queuing

You can add that on top of react-query (or just use it with fetch) to create one queue per record and enforce that they are only sent one after each other.

**Reply by jbojcic1** [reply]
> Maybe I’m misunderstand but why are you doing this on the frontend at all? It seems like synchronising two systems that should happen in the backend.
> 
> >

Normally yes but it's a specific app where the data needs to be accessible only to the client and thus is encrypted on the client side (with the key only client can read) before writing it to the Supabase db. 

Will check the pacer. Thx for suggestion.

**Comment by DamianOsipiuk** [comment]
@TkDodo Two things I would like to clarify.
1. I assume `scope` would be also available on `defaultOptions` to basically get serial mutations by default?
2. To make it work properly with `persisters`, we would need some kind of `order` property for stored mutations. Then when restoring them on app load, we would be able to properly queue them when same `scope` is assigned to many of them.

**Reply by TkDodo** [reply]
1) yes
2) the order is determined by the order in the array. When dehydrating, we call getAll(), which transforms the nested structure to a flat list, but the order is preserved. Then, when hydrating, we go through the list one by one, so the order should be the same.

**Comment by momme-rtf** [comment]
@TkDodo We would be quite interested in this enhancement. Do you already know if and when we can expect it to be implemented?

**Reply by ReangeloJ** [reply]
It was just released!

https://x.com/tkdodo/status/1782299195279393182?s=46&t=ylVAbPFr-Qat5giLV4s9qg

**Reply by TkDodo** [reply]
yes, it's shipped in https://github.com/TanStack/query/releases/tag/v5.31.0

**Reply by momme-rtf** [reply]
Awesome thanks!:)

**Comment by Kass86** [comment]
Can we use it now? @TkDodo

**Reply by TkDodo** [reply]
sure: https://tanstack.com/query/latest/docs/framework/react/guides/mutations#mutation-scopes

**Comment by boosja** [comment]
This version broke our app.

When the app goes online after being offline all mutations run in parallel. This breaks our app since we require serial execution after being offline. 

I guess I can add the same scope to every mutation we have in our app, but this feature is still a breaking change in the library.
- Version 5.29.2 runs the mutations in serial when going online. 
- Version 5.31.0 runs them in parallel as described above.

The feature itself is a very nice addition! An option to use scope in the mutate call is a great proposal IMO.

**Reply by TkDodo** [reply]
I'm aware that we changed the behaviour, but I would rather see it as a bugfix because the same mutations used to run in parallel when you are online, but in serial when you came online after having been offline. So clicking the same button twice behaved differently. Now, this bug is fixed because it will behave the same no matter if you are online when the mutation is triggered or if you are coming back from being offline. It's all dependent on the mutation scope.

The documentation also never said that they run in serial, just that they will be executed in the same order:

> If mutations fail because the device is offline, they will be retried in the same order when the device reconnects.

The fix is indeed to add the same `scope` for all mutations, or even to set the same scope globally for all mutations.

**Reply by boosja** [reply]
I see. Thanks for responding.

Is it possible to have them run in parallel while online, but serially after being offline?

**Reply by TkDodo** [reply]
I don't think so

**Comment by axelv** [comment]
Has the `takeLatest` flag been implemented?

**Reply by TkDodo** [reply]
nope, not yet, there hasn’t been much demand. do you want to look into it?

**Reply by axelv** [reply]
Still figuring out if we need this.
If so, I'm happy to look into it! This would be my first contribution, though.
